src/re_frame/frame.cljs

```diff
@@ -0,0 +1,175 @@
+    (fn [reducing-fn]                                       ; <- this is a transducer
+      (fn
+        ([] (reducing-fn))                                  ; transduction init, see [1]
+        ([result] (reducing-fn result))                     ; transduction completion, see [1]
+        ([state event]                                      ; transduction step, see [1]
+         (let [event-id (get-event-id event)
+               handler-fn (event-id (:handlers frame))]
+           (if (nil? handler-fn)
+             (do
+               (error frame "re-frame: no event handler registered for: \"" event-id "\". Ignoring.")
+               state)
+             (let [old-db (db-selector state)               ; db-selector is responsible for retrieving actual db from current state
+                   new-db (handler-fn old-db event)]        ; calls selected handler (including all composed middlewares)
+               (if (nil? new-db)                            ; TODO: this test should be optional, there could be valid use-cases for nil db
+                 (do
+                   (error frame "re-frame: your handler returned nil. It should return the new db state. Ignoring.")
+                   state)
+                 (reducing-fn state new-db))))))))))        ; reducing function prepares new transduction state
+
+(defn get-frame-transducer
+  "Returns a transducer: state, event -> state.
+This transducer resolves event-id, selects matching handler and calls it with old db state to produce a new db state.
+Transducer must have no knowledge of underlying app-db-atom, reagent, core.async or anything else out there.
+All business of queuing events and application of their effects must be baked into reducing-fn and provided from outside
+by the process doing actual transduction. See event processing helpers below for an example."
+  ([frame] (get-frame-transducer frame identity))
+  ([frame db-selector] ((frame-transducer-factory frame) db-selector)))
+
+;; -- event processing -------------------------------------------------------------------------------------------------
+
+(defn process-event [frame init-db event]
+  (let [reducing-fn (fn [_old-state new-state] new-state)
+        xform (get-frame-transducer frame identity)]
+    ((xform reducing-fn) init-db event)))
+
+(defn process-events [frame init-db events]
+  (let [reducing-fn (fn
+                      ([db-states] db-states)               ; completion
+                      ([db-states new-db]                   ; in each step
+                       (conj db-states new-db)))            ; add new-db state to the vector
+        xform (get-frame-transducer frame last)]
+    (transduce xform reducing-fn [init-db] events)))
+
+(defn process-event-on-atom [frame db-atom event]
+  (let [old-db @db-atom
+        new-db (process-event frame old-db event)]
+    (if-not (identical? old-db new-db)
+      (reset! db-atom new-db))))
```

Comment:
This pattern:
```clojure
(if-not (identitcal? old-db new-db)
  (reset! db-atom new-db))
```
is appearing in quite a few places, it might be better refactor it out to a separate fn:
```clojure
(defn reset-if-changed! [db-atom old-db new-db] (...))
```